File: backend/src/morado/schemas/test_suite.py
"""Test Suite Pydantic schemas.

This module provides schemas for test suite API request/response validation.
"""


from pydantic import BaseModel, ConfigDict, Field

from morado.schemas.common import PaginatedResponse, TimestampMixin, UUIDMixin


class TestSuiteBase(BaseModel):
    """测试套件基础schema

    包含测试套件的基本字段，用于创建和更新操作。
    """

    name: str = Field(min_length=1, max_length=200, description="套件名称")
    description: str | None = Field(default=None, description="套件描述")
    execution_order: str = Field(default="sequential", pattern="^(sequential|parallel)$", description="执行顺序")
    parallel_execution: bool = Field(default=False, description="是否并行执行")
    continue_on_failure: bool = Field(default=True, description="失败时是否继续")
    schedule_config: dict | None = Field(default=None, description="调度配置")
    is_scheduled: bool = Field(default=False, description="是否启用调度")
    environment: str = Field(default="test", max_length=20, description="执行环境")
    global_variables: dict | None = Field(default=None, description="全局变量")
    tags: list[str] | None = Field(default=None, description="标签")
    version: str = Field(default="1.0.0", max_length=20, description="版本号")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "name": "用户模块回归测试套件",
                "description": "包含所有用户相关功能的测试用例",
                "execution_order": "sequential",
                "parallel_execution": False,
                "continue_on_failure": True,
                "environment": "test",
                "tags": ["回归测试", "用户模块"],
                "version": "1.0.0"
            }
        }
    )


class TestSuiteCreate(TestSuiteBase):
    """创建测试套件schema

    用于创建新的测试套件。
    """

    created_by: int | None = Field(default=None, description="创建者ID")


class TestSuiteUpdate(BaseModel):
    """更新测试套件schema

    所有字段都是可选的，只更新提供的字段。
    """

    name: str | None = Field(default=None, min_length=1, max_length=200, description="套件名称")
    description: str | None = Field(default=None, description="套件描述")
    execution_order: str | None = Field(default=None, pattern="^(sequential|parallel)$", description="执行顺序")
    parallel_execution: bool | None = Field(default=None, description="是否并行执行")
    continue_on_failure: bool | None = Field(default=None, description="失败时是否继续")
    schedule_config: dict | None = Field(default=None, description="调度配置")
    is_scheduled: bool | None = Field(default=None, description="是否启用调度")
    environment: str | None = Field(default=None, max_length=20, description="执行环境")
    global_variables: dict | None = Field(default=None, description="全局变量")
    tags: list[str] | None = Field(default=None, description="标签")
    version: str | None = Field(default=None, max_length=20, description="版本号")

    model_config = ConfigDict(from_attributes=True)


class TestSuiteResponse(TestSuiteBase, TimestampMixin, UUIDMixin):
    """测试套件响应schema

    包含完整的测试套件信息，包括ID和时间戳。
    """

    id: int = Field(description="套件ID")
    created_by: int | None = Field(default=None, description="创建者ID")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "uuid": "550e8400-e29b-41d4-a716-446655440000",
                "name": "用户模块回归测试套件",
                "description": "包含所有用户相关功能的测试用例",
                "execution_order": "sequential",
                "parallel_execution": False,
                "continue_on_failure": True,
                "environment": "test",
                "tags": ["回归测试", "用户模块"],
                "version": "1.0.0",
                "created_by": 1,
                "created_at": "2024-01-01T00:00:00Z",
                "updated_at": "2024-01-01T00:00:00Z"
            }
        }
    )


class TestSuiteListResponse(PaginatedResponse[TestSuiteResponse]):
    """测试套件列表响应schema"""


# Test Suite Case Schemas

class TestSuiteCaseBase(BaseModel):
    """测试套件-用例关联基础schema"""

    test_suite_id: int = Field(description="测试套件ID")
    test_case_id: int = Field(description="测试用例ID")
    execution_order: int = Field(default=0, description="执行顺序")
    is_enabled: bool = Field(default=True, description="是否启用")
    case_parameters: dict | None = Field(default=None, description="用例参数覆盖")
    description: str | None = Field(default=None, description="说明")

    model_config = ConfigDict(from_attributes=True)


class TestSuiteCaseCreate(TestSuiteCaseBase):
    """创建测试套件-用例关联schema"""


class TestSuiteCaseResponse(TestSuiteCaseBase, TimestampMixin):
    """测试套件-用例关联响应schema"""

    id: int = Field(description="关联ID")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/test-suite.ts
/**
 * Test Suite API Client
 *
 * Test Suite Management and Execution
 * Provides functions for managing test suites that group multiple test cases.
 */

import apiClient, { type PaginatedResponse } from './index';

// Test Suite Types
export interface TestSuite {
  id: number;
  uuid: string;
  name: string;
  description?: string;
  execution_order: string;
  parallel_execution: boolean;
  continue_on_failure: boolean;
  schedule_config?: Record<string, unknown>;
  is_scheduled: boolean;
  environment: string;
  global_variables?: Record<string, unknown>;
  tags?: string[];
  version: string;
  created_by?: number;
  created_at: string;
  updated_at: string;
}

export interface TestSuiteCreate {
  name: string;
  description?: string;
  execution_order?: string;
  parallel_execution?: boolean;
  continue_on_failure?: boolean;
  schedule_config?: Record<string, unknown>;
  is_scheduled?: boolean;
  environment?: string;
  global_variables?: Record<string, unknown>;
  tags?: string[];
  version?: string;
  created_by?: number;
}

export interface TestSuiteUpdate {
  name?: string;
  description?: string;
  execution_order?: string;
  parallel_execution?: boolean;
  continue_on_failure?: boolean;
  schedule_config?: Record<string, unknown>;
  is_scheduled?: boolean;
  environment?: string;
  global_variables?: Record<string, unknown>;
  tags?: string[];
  version?: string;
}

export interface TestSuiteCase {
  id: number;
  test_suite_id: number;
  test_case_id: number;
  execution_order: number;
  is_enabled: boolean;
  case_parameters?: Record<string, unknown>;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface TestSuiteCaseCreate {
  test_suite_id: number;
  test_case_id: number;
  execution_order?: number;
  is_enabled?: boolean;
  case_parameters?: Record<string, unknown>;
  description?: string;
}

export interface TestSuiteCaseUpdate {
  execution_order?: number;
  is_enabled?: boolean;
  case_parameters?: Record<string, unknown>;
  description?: string;
}

export interface TestSuiteListParams {
  page?: number;
  page_size?: number;
  is_scheduled?: boolean;
  search?: string;
  tags?: string[];
  environment?: string;
}

export interface TestSuiteExecutionResult {
  success: boolean;
  test_suite_id: number;
  execution_id: number;
  test_case_results: Array<{
    test_case_id: number;
    test_case_name: string;
    success: boolean;
    duration: number;
    error?: string;
  }>;
  total_duration: number;
  timestamp: string;
  environment: string;
  passed_count: number;
  failed_count: number;
  skipped_count: number;
}

// Test Suite API Functions

/**
 * Get list of test suites with pagination and filtering
 */
export async function getTestSuites(
  params?: TestSuiteListParams,
): Promise<PaginatedResponse<TestSuite>> {
  const response = await apiClient.get<PaginatedResponse<TestSuite>>('/test-suites', { params });
  return response.data;
}

/**
 * Get a single test suite by ID
 */
export async function getTestSuite(id: number): Promise<TestSuite> {
  const response = await apiClient.get<TestSuite>(`/test-suites/${id}`);
  return response.data;
}

/**
 * Create a new test suite
 */
export async function createTestSuite(data: TestSuiteCreate): Promise<TestSuite> {
  const response = await apiClient.post<TestSuite>('/test-suites', data);
  return response.data;
}

/**
 * Update an existing test suite
 */
export async function updateTestSuite(id: number, data: TestSuiteUpdate): Promise<TestSuite> {
  const response = await apiClient.patch<TestSuite>(`/test-suites/${id}`, data);
  return response.data;
}

/**
 * Delete a test suite
 */
export async function deleteTestSuite(id: number): Promise<void> {
  await apiClient.delete(`/test-suites/${id}`);
}

/**
 * Duplicate a test suite
 */
export async function duplicateTestSuite(id: number, newName?: string): Promise<TestSuite> {
  const response = await apiClient.post<TestSuite>(`/test-suites/${id}/duplicate`, {
    name: newName,
  });
  return response.data;
}

/**
 * Execute a test suite
 */
export async function executeTestSuite(
  id: number,
  runtimeParams?: Record<string, unknown>,
): Promise<TestSuiteExecutionResult> {
  const response = await apiClient.post<TestSuiteExecutionResult>(`/test-suites/${id}/execute`, {
    runtime_parameters: runtimeParams,
  });
  return response.data;
}

/**
 * Get scheduled test suites
 */
export async function getScheduledTestSuites(): Promise<TestSuite[]> {
  const response = await apiClient.get<PaginatedResponse<TestSuite>>('/test-suites', {
    params: { is_scheduled: true, page_size: 100 },
  });
  return response.data.items;
}

/**
 * Search test suites by name or tags
 */
export async function searchTestSuites(query: string): Promise<TestSuite[]> {
  const response = await apiClient.get<PaginatedResponse<TestSuite>>('/test-suites', {
    params: { search: query, page_size: 50 },
  });
  return response.data.items;
}

// Test Suite Case Association Functions

/**
 * Get test cases associated with a test suite
 */
export async function getTestSuiteCases(testSuiteId: number): Promise<TestSuiteCase[]> {
  const response = await apiClient.get<PaginatedResponse<TestSuiteCase>>(
    `/test-suites/${testSuiteId}/cases`,
  );
  return response.data.items;
}

/**
 * Add a test case to a test suite
 */
export async function addCaseToTestSuite(data: TestSuiteCaseCreate): Promise<TestSuiteCase> {
  const response = await apiClient.post<TestSuiteCase>('/test-suite-cases', data);
  return response.data;
}

/**
 * Update a test suite-case association
 */
export async function updateTestSuiteCase(
  id: number,
  data: TestSuiteCaseUpdate,
): Promise<TestSuiteCase> {
  const response = await apiClient.patch<TestSuiteCase>(`/test-suite-cases/${id}`, data);
  return response.data;
}

/**
 * Remove a test case from a test suite
 */
export async function removeCaseFromTestSuite(id: number): Promise<void> {
  await apiClient.delete(`/test-suite-cases/${id}`);
}

/**
 * Reorder test cases in a test suite
 */
export async function reorderTestSuiteCases(
  testSuiteId: number,
  caseOrders: Array<{ test_case_id: number; execution_order: number }>,
): Promise<void> {
  await apiClient.post(`/test-suites/${testSuiteId}/cases/reorder`, { case_orders: caseOrders });
}

/**
 * Get test suite execution history
 */
export async function getTestSuiteExecutionHistory(
  testSuiteId: number,
  limit = 10,
): Promise<TestSuiteExecutionResult[]> {
  const response = await apiClient.get<{ items: TestSuiteExecutionResult[] }>(
    `/test-suites/${testSuiteId}/executions`,
    { params: { limit } },
  );
  return response.data.items;
}

/**
 * Schedule a test suite
 */
export async function scheduleTestSuite(
  id: number,
  scheduleConfig: Record<string, unknown>,
): Promise<TestSuite> {
  const response = await apiClient.post<TestSuite>(`/test-suites/${id}/schedule`, {
    schedule_config: scheduleConfig,
    is_scheduled: true,
  });
  return response.data;
}

/**
 * Unschedule a test suite
 */
export async function unscheduleTestSuite(id: number): Promise<TestSuite> {
  const response = await apiClient.post<TestSuite>(`/test-suites/${id}/unschedule`, {
    is_scheduled: false,
  });
  return response.data;
}
